File: src/org/codepulsar/lang/GlobalVariable.java
package org.codepulsar.lang;

import org.codepulsar.primitives.Primitive;
import org.codepulsar.primitives.PrimitiveType;

import java.util.HashMap;

public class GlobalVariable {
    HashMap<String, Global> variables;

    public GlobalVariable() {
        this.variables = new HashMap<>();
    }

    public void addVariable(String name, Primitive value, PrimitiveType type, boolean isInitialized, boolean isConstant) {
        this.variables.put(name, new Global(value, type, isInitialized, isConstant));
    }

    public void reassignVariable(String name, Primitive value) {
        Global variable = this.variables.get(name);

        this.variables.put(name, new Global(value, variable.getType(), true, variable.isConstant()));
    }

    public Primitive getValue(String name) {
        return this.variables.get(name).getValue();
    }

    public PrimitiveType getType(String name) {
        return this.variables.get(name).getType();
    }

    public boolean containsVariable(String name) {
        Global global = this.variables.get(name);
        return global != null;
    }

    record Global(Primitive value, PrimitiveType type, boolean isInitialized, boolean isConstant) {
        public Primitive getValue() {
            return this.value;
        }

        public PrimitiveType getType() {
            return this.type;
        }
    }
}

File: src/codepulsar/variable/GlobalVariable.cpp
#include "GlobalVariable.h"


Pulsar::GlobalVariable::GlobalVariable(Primitive* value, Pulsar::PrimitiveType type, bool initialized, bool constant) {
    this->value = value;
    this->type = type;
    this->initialized = initialized;
    this->constant = constant;
}

Pulsar::Primitive* Pulsar::GlobalVariable::getValue() {
    return this->value;
}

Pulsar::PrimitiveType Pulsar::GlobalVariable::getType() {
    return this->type;
}

bool Pulsar::GlobalVariable::isConstant() {
    return this->constant;
}

bool Pulsar::GlobalVariable::isInitialized() {
    return this->initialized;
}

void Pulsar::GlobalVariable::setInitialized() {
    this->initialized = true;
}
